### PR TITLE
[compiler] Add optional validation message to `@pattern` decorator.

### LINKED ISSUE #2718: [compiler] Optional description for `@pattern`
TypeSpec supports several validations, like `@pattern`. For most validations, for example: `@minValue`, the error message that should be given when validation fails is obvious. However, for `@pattern`, the regular expressions given are often difficult to understand, and we've found that there are some use cases for attaching an optional description to the pattern for readability and for use when surfacing validation errors to end users.

I propose augmenting the definition of `@pattern` like this:

```tsp
extern dec pattern(target: string | bytes | ModelProperty, pattern: valueof string, description?: valueof string);
```

The `description` can state what the pattern checks for. For example:

```tsp
@pattern("^[a-zA-Z0-9_]{3, 24}$", "must be between 3 and 24 alphanumeric characters and underscores")
id: string;
```

### PR DESCRIPTION
**Pull request by witemple-msft**
Closes #2718

This change adds support for an optional message that emitters may use to communicate the context of a pattern validation error.

I also added some baseline tests for `@pattern` since there were none in decorators.spec.ts.

### CODE REVIEW

File: packages/compiler/src/lib/decorators.ts [hunk 1]
```diff
@@ -417,12 +419,20 @@ export function $pattern(
   }
 
   context.program.stateMap(patternValuesKey).set(target, pattern);
+
+  if (validationMessage) {
+    context.program.stateMap(patternValidationMessageKey).set(target, validationMessage);
```
Comment (timotheeguerin) [inline packages/compiler/src/lib/decorators.ts:424]: Why not change the type and use the existing state key

Comment (witemple-msft) [reply packages/compiler/src/lib/decorators.ts:424]: Good call, I like the idea of doing getPatternData and having it return an object that we can extend in future.

**Review by timotheeguerin — APPROVED** [review]

File: packages/compiler/src/lib/decorators.ts [hunk 2]
```diff
@@ -405,21 +405,53 @@ export function getFormat(program: Program, target: Type): string | undefined {
 
 const patternValuesKey = createStateSymbol("patternValues");
 
+export interface PatternData {
+  pattern: string;
+  validationMessage?: string;
+}
```
Comment (timotheeguerin) [inline packages/compiler/src/lib/decorators.ts:411]:
```suggestion
export interface PatternData {
  readonly pattern: string;
  readonly validationMessage?: string;
}
```
thinking we should start making things readonly in general as people will write more and more decorators,emitters and stuff changing that data will result in very unexpected result

Comment (witemple-msft) [reply packages/compiler/src/lib/decorators.ts:411]: Good point. I'll integrate this suggestion. Do you have any thoughts about making AST node properties read-only after we finish them? We want to strongly discourage writing to the AST for similar reasons.
